third_party/nvfuser/csrc/multidevice/multidevice_runtime.cpp

```diff
@@ -0,0 +1,199 @@
+  // Set parameters inferred by auto scheduler.
+  if (maybe_scheduler_entry.has_value()) {
+    auto scheduler_entry = maybe_scheduler_entry.value();
+    // TODO: I don't remember what happened with index mode. Need to follow up.
+    // args.setIndexMode(scheduler_entry->indexMode());
+    // Set launch parameters with auto scheduler.
+    updateLaunchParamsFromScheduler(scheduler_entry, launch_params);
+  }
+
+  args.setDeviceIndex(device_index);
+  // Lower the fusion and compile the generated kernel.
+  executor_ptr->compileFusion(fusion_from_cluster.get(), args, launch_params);
+
+  return executor_ptr;
+}
+
+void MultiDeviceRuntime::handle(SendRecv* sr) {
+  auto sender_cluster = sr->in()->getCluster();
+  auto receiver_cluster = sr->out()->getCluster();
+
+  // check if current process receives
+  bool is_sender = shouldRun(sender_cluster);
+  // check if current process sends
+  bool is_receiver = shouldRun(receiver_cluster);
+
+  int sender_rank = sender_cluster->params().process_rank;
+  int receiver_rank = receiver_cluster->params().process_rank;
+
+  // Val corresponding to the sent tensor
+  auto val = sr->in()->getOriginalVal();
+
+  // container for the sent/received tensor
+  std::vector<at::Tensor> tensor = {val_to_ivalue_.at(val).toTensor()};
+
+  if (is_sender) {
+    // sending the tensor
+    process_cluster_->send(tensor, receiver_rank, 0);
+  }
+  if (is_receiver) {
+    // receiving the tensor
+    auto work = process_cluster_->recv(tensor, sender_rank, 0);
+    // wait for completion
+    while (!work->isCompleted());
+    // store the receive tensor
+    val_to_ivalue_[val] = (IValue)(tensor[0]);
+  }
+}
+
+void MultiDeviceRuntime::handle(AggregateExpr* aExpr) {
+  auto cluster = aExpr->getCluster();
+
+  // get the IValues corresponding to the cluster's input
+  std::vector<IValue> cluster_input;
+  std::transform(cluster->inputs().begin(),
+                 cluster->inputs().end(),
+                 std::back_inserter(cluster_input),
+                 [this](auto input_val) {return val_to_ivalue_[input_val]; });
+
+  // Run the lowering and compilation step if we haven't compiled yet.
+  if (!compiled_kernels_.count(cluster)) {
+    compiled_kernels_[cluster] = compileCluster(cluster, cluster_input);
+  }
+  auto& executor = compiled_kernels_[cluster];
+
+  // Launch kernel and record the kernel output into current context
+  std::vector<at::Tensor> outputs;
+  if (shouldRun(cluster)) {
+    // Use default launch parameters.
+    LaunchParams launch_params;
+    // If the kernel was auto-scheduled, we need to
+    //  pull the launch parameters from the scheduler.
+    auto scheduler_it = auto_scheduler_registry_.find(cluster);
+    if (scheduler_it != auto_scheduler_registry_.end()) {
+      updateLaunchParamsFromScheduler(
+          scheduler_it->second.get(), launch_params);
+    }
+    outputs = executor->runFusion(cluster_input, launch_params, {});
+  } else {
+    // Allocate space for kernel outputs.
```

Comment:
What was the reason of allocating outputs even when a cluster is not executed?

Reply:
This should be optimize in the future, but in the meantime, we do it this way for two reasons:
1) it could be that these output will be used as inputs of a subsequent kernel. In this case we will receive this tensor from a SendRecv and so we nee to allocate a placeholder for this tensor.
2) It is also a way to to infer the shape of subsequent kernel's I/O

Reply:
Thanks. Please add it to the code as a comment.